File: packages/python/src/opensms/resources/contacts.py
"""The ``contacts`` resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from .._serialize import path_id, prune
from ..models import Contact, Page
from ._base import Resource


class Contacts(Resource):
    """Address book. Accessed as ``client.contacts``."""

    def list(self, *, limit: Optional[int] = None, cursor: Optional[str] = None) -> Page[Contact]:
        """List contacts."""
        return self._page(self._t.request("GET", "/v1/contacts", query={"limit": limit, "cursor": cursor}))

    def create(
        self,
        *,
        e164: str,
        name: Optional[str] = None,
        attributes: Optional[Dict[str, Any]] = None,
        idempotency_key: Optional[str] = None,
    ) -> Contact:
        """Create a contact. A duplicate ``e164`` returns ``409``."""
        body = prune({"e164": e164, "name": name, "attributes": attributes})
        return self._t.request(
            "POST", "/v1/contacts", json_body=body, idempotent=True, idempotency_key=idempotency_key
        )

    def get(self, id: str) -> Contact:
        """Fetch a contact."""
        return self._t.request("GET", f"/v1/contacts/{path_id(id)}")

    def update(
        self,
        id: str,
        *,
        e164: Optional[str] = None,
        name: Optional[str] = None,
        attributes: Optional[Dict[str, Any]] = None,
    ) -> Contact:
        """Update the given fields; others are kept (PATCH)."""
        body = prune({"e164": e164, "name": name, "attributes": attributes})
        return self._t.request("PATCH", f"/v1/contacts/{path_id(id)}", json_body=body)

    def delete(self, id: str) -> None:
        """Delete a contact."""
        self._t.request("DELETE", f"/v1/contacts/{path_id(id)}")

File: packages/ruby/lib/opensms/resources/contacts.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +contacts+ resource. Accessed as +client.contacts+.
    class Contacts < Base
      FIELDS = %i[e164 name attributes].freeze

      # GET /v1/contacts -> Page<Contact>.
      # @param params [Hash] :limit, :cursor
      # @return [Opensms::Page]
      def list(params = nil, **kwargs)
        http_page("/v1/contacts", Models.build(merge(params, kwargs), %i[limit cursor]))
      end

      # POST /v1/contacts -> 201 Contact.
      # @param params [Hash] :e164 (required), :name, :attributes
      def create(params = nil, idempotency_key: nil, **kwargs)
        body = Models.build(merge(params, kwargs), FIELDS, required: %i[e164])
        @transport.request(:post, "/v1/contacts", body: body, idempotency_key: idem(idempotency_key))
      end

      # GET /v1/contacts/{id} -> Contact.
      def get(id)
        http_get(path("/v1/contacts", id!(id)))
      end

      # Partial update. PATCH /v1/contacts/{id} -> Contact.
      # @param params [Hash] :e164, :name, :attributes
      def update(id, params = nil, **kwargs)
        body = Models.build(merge(params, kwargs), FIELDS)
        @transport.request(:patch, path("/v1/contacts", id!(id)), body: body)
      end

      # DELETE /v1/contacts/{id} -> 204.
      # @return [nil]
      def delete(id)
        http_delete(path("/v1/contacts", id!(id)))
      end
    end
  end
end
